source/dub/commandline.d

```diff
@@ -126,155 +445,76 @@ int runDubCommandLine(string[] args)
-		args = args[0 .. app_args_idx];
-	}
-	args = args[1 .. $]; // strip the application name
-
-	// handle direct dub options
-	if (args.length) switch (args[0])
-	{
-	case "--version":
-		showVersion();
-		return 0;
-
-	default:
-		break;
-	}
-
-	// parse general options
-	CommonOptions options;
-	LogLevel loglevel = LogLevel.info;
-
-	auto common_args = new CommandArgs(args);
-	try {
-		options.prepare(common_args);
+	auto common_args = new CommandArgs(args[1..$]);
 
-		if (options.vverbose) loglevel = LogLevel.debug_;
-		else if (options.verbose) loglevel = LogLevel.diagnostic;
-		else if (options.vquiet) loglevel = LogLevel.none;
-		else if (options.quiet) loglevel = LogLevel.warn;
-		else if (options.verror) loglevel = LogLevel.error;
-		setLogLevel(loglevel);
-	} catch (Throwable e) {
+	try handler.prepareOptions(common_args);
+	catch (Throwable e) {
 		logError("Error processing arguments: %s", e.msg);
 		logDiagnostic("Full exception: %s", e.toString().sanitize);
 		logInfo("Run 'dub help' for usage information.");
 		return 1;
 	}
 
-	if (options.root_path.empty)
-		options.root_path = getcwd();
-	else
+	if (handler.options.version_)
 	{
-		import std.path : absolutePath, buildNormalizedPath;
-
-		options.root_path = options.root_path.absolutePath.buildNormalizedPath;
+		showVersion();
+		return 0;
 	}
 
 	// extract the command
-	string cmdname;
-	args = common_args.extractRemainingArgs();
-	if (args.length >= 1 && !args[0].startsWith("-")) {
-		cmdname = args[0];
-		args = args[1 .. $];
-	} else {
-		if (options.help) {
-			showHelp(commands, common_args);
-			return 0;
-		}
-		cmdname = "run";
-	}
-	auto command_args = new CommandArgs(args);
+	args = common_args.extractAllRemainingArgs();
 
-	if (cmdname == "help") {
-		showHelp(commands, common_args);
-		return 0;
-	}
+	auto command_name_argument = CommandNameArgument(args);
 
-	// find the selected command
+	auto command_args = new CommandArgs(command_name_argument.extractAllRemainingArgs);
 	Command cmd;
-	foreach (grp; commands)
-		foreach (c; grp.commands)
-			if (c.name == cmdname) {
-				cmd = c;
-				break;
-			}
 
-	if (!cmd) {
-		logError("Unknown command: %s", cmdname);
-		writeln();
-		showHelp(commands, common_args);
-		return 1;
-	}
-
-	// process command line options for the selected command
 	try {
-		cmd.prepare(command_args);
-		enforceUsage(cmd.acceptsAppArgs || app_args.length == 0, cmd.name ~ " doesn't accept application arguments.");
+		cmd = handler.prepareCommand(command_name_argument.value, command_args);
 	} catch (Throwable e) {
 		logError("Error processing arguments: %s", e.msg);
 		logDiagnostic("Full exception: %s", e.toString().sanitize);
 		logInfo("Run 'dub help' for usage information.");
 		return 1;
 	}
 
-	if (options.help) {
+	if (cmd is null) {
+		logError("Unknown command: %s", command_name_argument.value);
+		writeln();
+		showHelp(handler.commands, common_args);
+		return 1;
+	}
+
+	if (cmd.name == "help") {
```

Comment:
`cast(HelpCommand)cmd` better describes the relation between this check and HelpCommand